discovery-infra/oc_utils.py

```diff
@@ -0,0 +1,169 @@
+import os
+import yaml
+
+import consts
+import requests
+import openshift as oc
+
+from retry import retry
+from logger import log
+
+
+def add_oc_user_arguments(parser):
+    parser.add_argument(
+        '--oc-mode',
+        help='If set, use oc instead of minikube',
+        action='store_true',
+        default=False
+    )
+    parser.add_argument(
+        '-ocu',
+        '--oc-user',
+        help='Username for oc login',
+        type=str
+    )
+    parser.add_argument(
+        '-ocp',
+        '--oc-pass',
+        help='Password for oc login',
+        type=str
+    )
+    parser.add_argument(
+        '-oct',
+        '--oc-token',
+        help='Token for oc login (an alternative for --oc-user & --oc-pass)',
+        type=str,
+        default='http'
+    )
+    parser.add_argument(
+        '-ocs',
+        '--oc-server',
+        help='Server for oc login, required if --oc-token is provided',
+        type=str,
+        default='https://api.ocp.prod.psi.redhat.com:6443'
+    )
+    parser.add_argument(
+        '--oc-scheme',
+        help='Scheme for assisted-service url on oc',
+        type=str,
+        default='http'
+    )
+
+
+def oc_login(**kwargs):
+    log.info('Performing oc login')
+    r = oc.Result('login')
+    r.add_action(
+        oc.oc_action(
+            context=oc.cur_context(),
+            verb='login',
+            cmd_args=_get_login_args_from_kwargs(kwargs)
+        )
+    )
+    r.fail_if('Error when trying to login')
+
+
+def _get_login_args_from_kwargs(kwargs):
+    if kwargs.get('username') and kwargs.get('password'):
+        return [
+            '--username', kwargs['username'],
+            '--password', kwargs['password'],
+            '--insecure-skip-tls-verify'
+        ]
+    elif kwargs.get('token') and kwargs.get('server'):
+        return [
+            '--token', kwargs['token'],
+            '--server', kwargs['server'],
+            '--insecure-skip-tls-verify'
+        ]
+    token, server = _get_token_and_server_from_kubeconfig()
+    return [
+        '--token', token,
+        '--server', server,
+        '--insecure-skip-tls-verify'
+    ]
+
+
+def _get_token_and_server_from_kubeconfig():
+    config = _load_kubeconfig()
+    for cluster in _iterate_clusters(config):
+        server = _get_server_from_cluster(cluster)
```

Comment:
A more pythonic way would be 
```python
    for cluster in _iterate_clusters(config):
         try:
             server = _get_server_from_cluster(cluster)
             user = _get_user_by_cluster(config, cluster)
             token = _get_token_by_user(config, user)
             return token, server
        except KeyError:
            continue
...

def _get_server_from_cluster(cluster):
    if f'.{consts.OC_DOMAIN}' not in cluster['cluster']['server']:
        raise KeyError(f'.{consts.OC_DOMAIN}')

    return cluster['cluster']['server']
```

Reply:
I don't agree so much. 
each function does it's own checks and returns nothing if failed to find what is was looking for.
does not necessarily KeyError, so it will force me to raise exceptions which is less elegant why and also not truth in all the cases. Even in your example you raised key error (where is should be value error instead).